worker/modelcache/worker.go

```diff
@@ -273,11 +275,40 @@ func (c *cacheWorker) translate(d multiwatcher.Delta) interface{} {
 			MachineId:      value.MachineId,
 			Ports:          coreNetworkPorts(value.Ports),
 			PortRanges:     coreNetworkPortRanges(value.PortRanges),
+			Principal:      value.Principal,
 			Subordinate:    value.Subordinate,
 			WorkloadStatus: coreStatus(value.WorkloadStatus),
 			AgentStatus:    coreStatus(value.AgentStatus),
 		}
-
+	case "charm":
+		if d.Removed {
+			return cache.RemoveCharm{
+				ModelUUID: id.ModelUUID,
+				CharmURL:  id.Id,
+			}
+		}
+		value, ok := d.Entity.(*multiwatcher.CharmInfo)
+		if !ok {
+			c.config.Logger.Errorf("unexpected type %T", d.Entity)
+			return nil
+		}
+		charmChange := cache.CharmChange{
+			ModelUUID: value.ModelUUID,
+			CharmURL:  value.CharmURL,
+		}
+		profile := value.LXDProfile
+		if profile != nil {
+			charmChange.LXDProfiler = lxdprofile.NewLXDCharmProfiler(
+				lxdprofile.Profile{
+					Config:      profile.Config,
+					Description: profile.Description,
+					Devices:     profile.Devices,
+				},
+			)
+		} else {
+			charmChange.LXDProfiler = lxdprofile.NewLXDCharmProfiler(lxdprofile.Profile{})
```

Comment:
Is this so we don't have to deal with `nil` as much?

Reply:
yes